File: android/src/main/java/com/jvanila/flutter/shopify/usecases/EditCustomerAddressUseCase.java
package com.jvanila.flutter.shopify.usecases;

import com.jvanila.flutter.pluginarch.PluginContext;
import com.shopapp.gateway.ApiCallback;
import com.shopapp.gateway.entity.Address;
import com.shopapp.gateway.entity.Error;
import com.shopapp.shopify.api.ShopifyApi;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import kotlin.Unit;

public class EditCustomerAddressUseCase extends ShopifyCallUseCase {

    private static final String ARG_ADDRESS_ID = "addressId";
    private static final String ARG_PRIMARY_ADDRESS = "primaryAddress";
    private static final String ARG_SECOND_ADDRESS = "secondAddress";
    private static final String ARG_CITY = "city";
    private static final String ARG_STATE = "state";
    private static final String ARG_COUNTRY = "country";
    private static final String ARG_ZIP = "zip";
    private static final String ARG_FIRST_NAME = "firstName";
    private static final String ARG_LAST_NAME = "lastName";
    private static final String ARG_COMPANY = "company";
    private static final String ARG_PHONE = "phone";

    public EditCustomerAddressUseCase(PluginContext<ShopifyApi> pluginContext) {
        super(pluginContext);
    }

    @Override
    protected void call(MethodCall input, final MethodChannel.Result result) {
        String id = input.argument(ARG_ADDRESS_ID);
        String primaryAddress = input.argument(ARG_PRIMARY_ADDRESS);
        String secondAddress = input.argument(ARG_SECOND_ADDRESS);
        String city = input.argument(ARG_CITY);
        String state = input.argument(ARG_STATE);
        String country = input.argument(ARG_COUNTRY);
        String firstName = input.argument(ARG_FIRST_NAME);
        String lastName = input.argument(ARG_LAST_NAME);
        String zip = input.argument(ARG_ZIP);
        String company = input.argument(ARG_COMPANY);
        String phone = input.argument(ARG_PHONE);

        Address address = new Address(id, primaryAddress, secondAddress, city, country,
                state, firstName, lastName, zip, company, phone);

        mPluginContext.api.instance.editCustomerAddress(id, address, new ApiCallback<Unit>() {
            @Override
            public void onResult(Unit unit) {
                result.success(true);
            }

            @Override
            public void onFailure(Error error) {
                System.out.println("onFailure -- " + error);
                result.error("EditCustomerAddressUseCase", error.getMessage(), error);
            }
        });
    }
}

File: ios/Classes/UseCases/EditCustomerAddressUseCase.swift
//
//  EditCustomerAddressUseCase.swift
//  shopify
//
//  Created by Joseph jose on 10/09/19.
//

import UIKit
import ShopApp_Shopify
import ShopApp_Gateway
import Flutter

class EditCustomerAddressUseCase: UseCase {

    static let ARG_ADDRESS_ID = "addressId"
    static let ARG_PRIMARY_ADDRESS = "primaryAddress"
    static let ARG_SECOND_ADDRESS = "secondAddress"
    static let ARG_CITY = "city"
    static let ARG_STATE = "state"
    static let ARG_COUNTRY = "country"
    static let ARG_ZIP = "zip"
    static let ARG_FIRST_NAME = "firstName"
    static let ARG_LAST_NAME = "lastName"
    static let ARG_COMPANY = "company"
    static let ARG_PHONE = "phone"
    
    override public init(_ context: PluginContext) {
        super.init(context)
    }
    
    override public func trigger(with methodCall: FlutterMethodCall, result: @escaping (Any?) -> Void) {
        
        if let args = methodCall.arguments as? [String:String] {
            let id = args[EditCustomerAddressUseCase.ARG_ADDRESS_ID]
            let primaryAddress = args[EditCustomerAddressUseCase.ARG_PRIMARY_ADDRESS]
            let secondAddress = args[EditCustomerAddressUseCase.ARG_SECOND_ADDRESS]
            let city = args[EditCustomerAddressUseCase.ARG_CITY]
            let state = args[EditCustomerAddressUseCase.ARG_STATE]
            let country = args[EditCustomerAddressUseCase.ARG_COUNTRY]
            let firstName = args[EditCustomerAddressUseCase.ARG_FIRST_NAME]
            let lastName = args[EditCustomerAddressUseCase.ARG_LAST_NAME]
            let zip = args[EditCustomerAddressUseCase.ARG_ZIP]
            let company = args[EditCustomerAddressUseCase.ARG_COMPANY]
            let phone = args[EditCustomerAddressUseCase.ARG_PHONE]
            
            let address = Address()
            address.firstName = firstName
            address.lastName = lastName
            address.address = primaryAddress
            address.secondAddress = secondAddress
            address.city = city
            address.state = state
            address.country = country
            address.zip = zip
            address.phone = phone
            address.id = id ?? ""
            
            (mContext.api.instance as! ShopifyAPI).updateCustomerAddress(with: address) { (success, error) in
                
                if error != nil {
                    
                    self.createError(withCase: "EditCustomerAddressUseCase", message: error!.errorMessage, error: error!, on: result)
                }else {
                    
                    result(success)
                }
            }
            
        }
    }
    
}
